lib/ui/flutter_dart_api.cc

```diff
@@ -0,0 +1,53 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/flutter_dart_api.h"
+
+#include <cstdio>
+
+#ifdef _DEBUG
+// Needed by dartutils.h below.
+#define DEBUG
+#endif  // _DEBUG
+
+#include "third_party/dart/runtime/bin/dartutils.h"
+#include "topaz/lib/tonic/converter/dart_converter.h"
+
+namespace blink {
+
+static Dart_Handle ExceptionValueForType(DartErrorType type,
+                                         const char* format,
+                                         va_list message_args) {
+  char message_buffer[256];
+  vsnprintf(message_buffer, sizeof(message_buffer), format, message_args);
+  switch (type) {
+    case DartErrorType::kArgumentError:
+      return ::dart::bin::DartUtils::NewDartArgumentError(message_buffer);
+    case DartErrorType::kEngineError:
+      // TODO(chinmaygarde): Placeholder till the engine gets its own error
+      // type.
```

Comment:
we can use StateError for most things, i imagine

Reply:
and Exception for other things.

There's a difference between "Error" and "Exception" in Dart, by the way. Errors are for things that indicate the code is bad -- e.g. you passed in a null, or you used a disposed object. Exceptions are for runtime problems that could not have been avoided by better code, e.g. being out of memory, the GPU not supporting something, etc.